pkg/server/chaosd/kafka.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2020 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chaosd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/magiconair/properties"
+	"github.com/pingcap/log"
+	perr "github.com/pkg/errors"
+	client "github.com/segmentio/kafka-go"
+	"github.com/segmentio/kafka-go/sasl/scram"
+	"go.uber.org/zap"
+
+	"github.com/chaos-mesh/chaosd/pkg/core"
+)
+
+type kafkaAttack struct{}
+
+var KafkaAttack AttackType = kafkaAttack{}
+
+func (j kafkaAttack) Attack(options core.AttackConfig, env Environment) (err error) {
+	attack := options.(*core.KafkaCommand)
+	switch attack.Action {
+	case core.KafkaFillAction:
+		return attackKafkaFill(attack)
+	case core.KafkaFloodAction:
+		return attackKafkaFlood(context.TODO(), attack)
+	case core.KafkaIOAction:
+		return attackKafkaIO(attack)
+	default:
+		return nil
+	}
+}
+
+func (j kafkaAttack) Recover(exp core.Experiment, env Environment) error {
+	attack := new(core.KafkaCommand)
+	if err := json.Unmarshal([]byte(exp.RecoverCommand), attack); err != nil {
+		return perr.Wrap(err, "unmarshal kafka command")
+	}
+	if attack.Action == core.KafkaIOAction {
+		return recoverKafkaIO(attack, env)
+	}
+	return nil
+}
+
+func dial(attack *core.KafkaCommand) (conn *client.Conn, err error) {
+	dialer := &client.Dialer{
+		Timeout:   10 * time.Second,
+		DualStack: true,
+	}
+	if attack.Username != "" {
+		dialer.SASLMechanism, err = scram.Mechanism(scram.SHA512, attack.Username, attack.Password)
+		if err != nil {
+			return nil, perr.Wrap(err, "create scram mechanism")
+		}
+	}
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	conn, err = dialer.DialLeader(context.Background(), "tcp", endpoint, attack.Topic, attack.Partition)
+	if err != nil {
+		return nil, perr.Wrapf(err, "dial kafka leader %s", endpoint)
+	}
+	return conn, nil
+}
+
+func attackKafkaFill(attack *core.KafkaCommand) (err error) {
+	// TODO: make it configurable
+	const messagePerRequest = 128
```

Comment:
seems that it is already configurable

Reply:
No, it's non-configurable. The `messagePerRequest` means the number of messages in a single request.